include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -65,6 +65,52 @@ class tracking_resource_adaptor final : public device_memory_resource {
         allocation_size{size} {};
   };
 
+  /**
+   * @brief Stores the current, peak, and total number of allocations and
+   * allocated bytes
+   */
+  struct allocation_counts {
```

Comment:
There are two things in this class (bytes and counts), that get updated in exactly the same way. I think it would be DRYer to make one simpler counter class and use it twice.

```suggestion
  struct counter {
    int64_t value{0};
    int64_t peak{0};
    int64_t total{0};
    
    counter& operator+=(int64_t x) { 
      value += x;
      total += x;
      peak = std::max(value, peak);
      return *this;
    }
        
    counter& operator-=(int64_t x) { 
      value -= x; 
      return *this;
    }
  };
```

In the `tracking_resource_adaptor` class, you just have two of these as members:

```
counter allocation_bytes;
counter allocation_count;
```

Then, in `do_allocate()`, just do 

```
allocation_bytes += bytes;
allocation_count += 1;
```

`do_deallocate()`:

```
allocation_bytes -= bytes;
allocation_count -= 1;
```

And get rid of `get_allocation_counts` in favor of a `get_counter` which lets you get one or the other...

Reply:
Like the idea. Resolving.